### DISCUSSION TITLE
SSL handshake timed out and channel closed exceptions

### DISCUSSION THREAD

**Discussion by SG87** [fastapi/fastapi#6801] [Questions] [answered]
The simplified code
```
import os
from elasticsearch import AsyncElasticsearch
from fastapi import FastAPI, HTTPException, Depends, status
import secrets
from fastapi.security import HTTPBasic, HTTPBasicCredentials
import uvicorn

application = FastAPI()
security = HTTPBasic()

es_client = AsyncElasticsearch([os.getenv('ES_URI')] , maxsize=10000)
es_index = os.getenv('ES_INDEX')
es_doc_type = os.getenv('ES_DOC_TYPE')

def authorize(credentials: HTTPBasicCredentials = Depends(security)):
    correct_username = secrets.compare_digest(credentials.username, os.getenv("BASIC_AUTH_USERNAME"))
    correct_password = secrets.compare_digest(credentials.password, os.getenv("BASIC_AUTH_PASSWORD"))
    if not (correct_username and correct_password):
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Incorrect username or password",
            headers={"WWW-Authenticate": "Basic"},
        )
    return True

query = {
    'query': {
        'function_score': {
            'query': {
                         'bool': ...
                     }
            },
        'score_mode': 'sum', 'boost_mode': 'sum'
    }, 
    'size': 6, 
    '_source': [...]
}


@application.get("/api/v1/smart-search", dependencies=[Depends(authorize)])
async def search(size:int=12, model:str=None, carSize:int=None): 
    ...
    es_res = await es_client.search(index=es_index, doc_type=es_doc_type, body=  query, max_concurrent_shard_requests=5000)
    return es_res

if __name__ == "__main__":
    uvicorn.run(application, host="0.0.0.0", port=5000)
```

### Description
The above code is simplified and creates a dynamic ES query.  Executes the ES query using AsyncElasticsearch. The result is parsed and returned. I tried multiple configuration for amongst others: 

- maxsize=10000 in AsyncElasticsearch client definition
- max_concurrent_shard_requests in search
- etc.

The application runs in docker. This is the dockerfile CMD:

`CMD gunicorn --do-handshake-on-connect --worker-class uvicorn.workers.UvicornWorker --worker-connections 1000 --workers 3 --bind 0.0.0.0:5000 api`

I tried running with:

- uvicorn, gunicorn with gevent and gunicorn with uvicorn worker as worker class. 
- Different number of workers, worker-connections, etc.
- Without and with do-handshake-on-connect

All this runs fine for:
- Local setup (Run API local in docker with local ES instance) with as many concurrent users as wanted
- API running in Kubernetes environment with (2 to 5 pods) and AWS c5.xlarge.elasticsearch ES cluster up to **60 concurrent users**

The API starts failing when running in Kubernetes environment with (2 to 5 pods) and AWS c5.xlarge.elasticsearch ES cluster starting for about **70 concurrent users or more**. Load generated on API (pods) is not too high and load on ES is low.

Then 70% of the calls fails with the mainly following errors:

- SSLException: handshake timed out (59%)
- ClosedChannelException (36%)

I don't have more information. Because this is all that is by Gatling load testing tool.

### Environment

requirements.txt

```
elasticsearch[async]>=7.12.0
starlette==0.13.6
elastic-apm==5.8.1
fastapi==0.60.1
uvicorn==0.11.8
gunicorn==20.1.0
```

* Python version:

Version: 3.6

`FROM python:3.6-alpine
`

**Comment by ycd** [comment]
All the low level networking stuff handled by either uvicorn or gunicorn (Layer 4), FastAPI runs on Layer 7 and doesn't knows anything about the layers below including things like SSL, TLS, Sockets, so this is not a FastAPI problem, you'll have a better chance to get an answer in those repositories.

**Comment by SG87** [comment, marked as answer]
Ok, thanks. I have reposted this in the uvicorn repository.

**Reply by Kludex** [reply]
Reference: https://github.com/encode/uvicorn/issues/1021

**Comment by tiangolo** [comment]
Thanks for the help here @ycd ! 👏  🙇 

As this wouldn't be related to FastAPI itself, I'm gonna close this one now.

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
